earth2studio/statistics/rank.py

```diff
@@ -0,0 +1,158 @@
+        return "rank_histogram"
+
+    @property
+    def reduction_dimensions(self) -> list[str]:
+        return [self.ensemble_dimension] + self._reduction_dimensions
+
+    def __call__(
+        self,
+        x: torch.Tensor,
+        x_coords: CoordSystem,
+        y: torch.Tensor,
+        y_coords: CoordSystem,
+    ) -> tuple[torch.Tensor, CoordSystem]:
+        """
+        Apply metric to data `x` and `y`, checking that their coordinates
+        are broadcastable. While reducing over `reduction_dimensions`.
+
+        Parameters
+        ----------
+        x : torch.Tensor
+            Input tensor of ensemble data. The rank of observation input tensor `y`
+            is determined with respect to the ensemble dimension of `x`.
+        x_coords : CoordSystem
+            Ordered dict representing coordinate system that describes the `x` tensor.
+            `reduction_dimensions` must be in coords.
+        y : torch.Tensor
+            The observation input tensor.
+        y_coords : CoordSystem
+            Ordered dict representing coordinate system that describes the `y` tensor.
+            `reduction_dimensions` must be in coords.
+
+        Returns
+        -------
+        tuple[torch.Tensor, CoordSystem]
+            Returns rank histogram tensor with appropriate reduced coordinates.
+        """
+        if not all([rd in x_coords for rd in self.reduction_dimensions]):
+            raise ValueError(
+                "Initialized reduction dimension does not appear in passed coords"
+            )
+
+        # Do some coordinate checking
+        # Assume ensemble_dim is in x_coords but not y_coords
+        if self.ensemble_dimension in y_coords:
+            raise ValueError(
+                f"{self.ensemble_dimension} should not be in y_coords but is."
+            )
+        if x.ndim != y.ndim + 1:
+            raise ValueError(
+                "x and y must have broadcastable shapes but got"
+                + f"{x.shape} and {y.shape}"
+            )
+        # Input coordinate checking
+        coord_count = 0
+        for c in x_coords:
+            if c != self.ensemble_dimension:
+                handshake_dim(y_coords, c, coord_count)
+                handshake_coords(y_coords, x_coords, c)
+                coord_count += 1
+
+        dim = list(x_coords).index(self.ensemble_dimension)
+
+        _ranks = torch.mean(1.0 * torch.ge(y, x), dim=dim)
+
+        # Reshape ranks using reduction dimensions
+        dims = [list(y_coords).index(rd) for rd in self._reduction_dimensions]
+        new_dims = list(range(len(dims)))
+        _ranks = torch.flatten(
+            torch.movedim(_ranks, dims, new_dims),
+            start_dim=new_dims[0],
+            end_dim=new_dims[-1],
+        )
+
+        # Compute histogram
+        _bin_edges = linspace(
+            torch.zeros_like(_ranks[0]), torch.ones_like(_ranks[0]), self.number_of_bins
+        )
+        _rank_histogram = _count_bins(_ranks, _bin_edges)
```

Comment:
Whats the expected dims of this? My understanding is [n_ensemble, (input_dims-ensemble_dims)...]

Reply:
Let's assume that `ensemble_dimension` is `ensemble` and `reduction_dimensions` is `[lat, lon]`. Then `x` starts life as, for the sake of argument, `[ensemble, time, variable, lat, lon]` and `y` needs to be `[time, variable, lat, lon]`. The ranking procedure reduces `x` to the same coordinates as `y`. Then we take the reduction dimensions and merge them together to take the histogram, so the input to `_count_bins` is `[ latxlon, time, variable]` and we return `[data, time, variable]`.

Reply:
In this case is it assumed that the reduction points are statistically independent?

What is `data` size?

Reply:
Oh I think its `bin` out from this function. Not data

Reply:
I guess we collect the bin centers and bin counts into a coordinate called "data" (there is probably a better name for this) and the bins are in a coordinate called "bin".